### PR TITLE
Support codepoints outside the BMP

### PR DESCRIPTION
**Pull request by seishun**

### CODE REVIEW

File: src/google/protobuf/compiler/js/js_generator.cc [hunk 1]
```diff
@@ -593,8 +593,8 @@ std::string JSOneofIndex(const OneofDescriptor* oneof) {
   return StrCat(index);
 }
 
-// Decodes a codepoint in \x0000 -- \xFFFF.
-uint16 DecodeUTF8Codepoint(uint8* bytes, size_t* length) {
+// Decodes a codepoint in \x0000 -- \x10FFFF.
+uint32 DecodeUTF8Codepoint(uint8* bytes, size_t* length) {
```
Comment (elharo) [inline src/google/protobuf/compiler/js/js_generator.cc:597]: I'm not a JavaScript expert but this seems unlikely to be backwards compatible. Can it be made so?

**Review by elharo — CHANGES_REQUESTED** [review]

File: src/google/protobuf/compiler/js/js_generator.cc (hunk 1 above)
Comment (seishun) [reply src/google/protobuf/compiler/js/js_generator.cc:597]: Backwards compatible in what way specifically?

Comment (elharo) [reply src/google/protobuf/compiler/js/js_generator.cc:597]: This changes the return type. Code that was expecting a uint16 is now getting a uint32 instead. Is that a problem in context? I'm  not sure, but I'm no JavaScript expert.

Comment (seishun) [reply src/google/protobuf/compiler/js/js_generator.cc:597]: This is C++, not JavaScript. This function is called only from `EscapeJSString`, which is updated accordingly.

Comment (elharo) [reply src/google/protobuf/compiler/js/js_generator.cc:597]: should be OK then

**Review by elharo — COMMENTED** [review]
Change looks good, but it needs some tests

**Comment by seishun** [comment]
> 
> 
> Change looks good, but it needs some tests

As far as I can tell, there are no tests for compiler output.

**Comment by acozzette** [comment]
We are closing all open JavaScript pull requests because the JavaScript implementation is moving to a new repository: https://github.com/protocolbuffers/protobuf-javascript Sorry for the inconvenience but please feel free to reopen the pull request in the new repo.

**Pull request closed without merging**